Add re::optimization()

hvds opened the pull request:
Given a compiled regexp object, this returns a hashref of the optimization
information discovered for it.

Docs and tests now added, I think it should be good to go.

Ref https://github.com/Perl/perl5/issues/18144

hvds committed:
Add re::optimization()

Given a compiled regexp object, this returns a hashref of the optimization
information discovered for it.

hvds commented on ext/re/re.xs:
```diff
@@ -143,3 +146,89 @@ PPCODE:
+    hv_stores(hv, "skip", _newSVbool(r->intflags & PREGf_SKIP));
+    hv_stores(hv, "implicit", _newSVbool(r->intflags & PREGf_IMPLICIT));
+
+    mPUSHp("stclass", 7);
```
oops, this line should have been removed, residue of an earlier attempt.

atoomic reviewed (commented):
It would be nice to have a few tests using `optimization` and checking its output?

hvds commented:
> It would be nice to have a few tests using `optimization` and checking its output?

Yes, that and docs - I plan to add those if people are happy with the general approach.

khwilliamson commented on ext/re/re.xs:
```diff
@@ -100,6 +101,8 @@ const struct regexp_engine wild_reg_engine = {
         my_re_op_compile,
 };
 
+#define _newSVbool(x) newSViv((x) ? 1 : 0)
```
It turns out that a leading underscore in something visible to C is illegal, reserved for the implementation.  We need to change ours to trailing underscores

hvds replied:
Thanks, I never heard of that; fixed.

khwilliamson commented on embed.fnc:
```diff
@@ -2320,9 +2320,16 @@ Ep	|SV*	|get_re_gclass_nonbitmap_data				   \
 #endif
-Ep	|void	|regprop	|NULLOK const regexp *prog|NN SV* sv|NN const regnode* o|NULLOK const regmatch_info *reginfo \
+#if defined(PERL_IN_REGCOMP_C) || defined(PERL_IN_REGEXEC_C) \
+ || defined(EXT_RE_RE_XS)
```
Did you check to see if the existing defined(PERL_EXT_RE_BUILD) works?

khwilliamson replied:
I kinda doubt that it does.

hvds replied:
I was unaware of it, I'll give it a go.

hvds replied:
Apparently it does work, I've dropped the extra symbol.

khwilliamson reviewed (commented):
This looks like a sane approach to me

tonycoz commented:
It looks reasonable to me.

hvds commented:
Thanks all, I'll try to finish it up and put it back in for formal review.

hvds committed:
reorder re.pm docs

regmust() is about internals, it should go at the end; move docs for
functions interrogating match results above it.

hvds commented:
Ugh, test fails on some platforms, looking into it.

hvds commented:
I'll withdraw this until I sort out the test failure.

hvds commented:
@khwilliamson (and anyone else) I need some opinion here.

The cause of the test failure (in at least some environments) is that Perl_regprop() does nothing on a non-DEBUGGING build, so when I call it to unwrap the stclass value I'm getting back undef. That could be avoided by calling my_regprop, the debug-enabled version from ext/re/re_comp.c.

The standard way to get that to happen automatically would be to include ext/re/re_top.h to get the appropriate defines, but doing that would break the recently added declaration of `wild_reg_engine` which has a mix of `Perl_*` and `my_*` functions.

So I have a few choices:
- include re_top.h further down in re.xs
  - I think this is a bad idea, could cause horrendous confusion for future readers (including me)
- invoke my_regprop directly
  - feels unclean, but might be ok; I think it'd be the first occurrence of any such reference outside the engine declarations
- move the declaration of the hybrid wild_reg_engine somewhere else, perhaps to its own file
  - this might be sensible regardless
- remove the #ifdef DEBUGGING protection from regprop, let the callers protect themselves instead
  - this might be sensible regardless, need to check all current callers in case there's some really good reason for the current state of affairs

hvds commented:
>  remove the #ifdef DEBUGGING protection from regprop, let the callers protect themselves instead
>       * this might be sensible regardless, need to check all current callers in case there's some really
>         good reason for the current state of affairs

All the current callers are themselves protected by #ifdef DEBUGGING, so I expected this to be simple. Sadly it is not - we're also relying on the fact that perl.h has an explicit `#define DEBUG_r(x) x` bypass for ext/re. Since regprop()'s body itself has a DEBUG_foo use within it, we rely on calling the ext/re version of it even within a debugging perl.

I'm not quite sure if this is all too clever by half, but I don't have a better solution for it right now.

hvds commented:
>     * invoke my_regprop directly
>       * feels unclean, but might be ok; I think it'd be the first occurrence of any such reference outside the engine declarations

I've gone for this pro tem, which also meant I no longer needed to change the exposure of Perl_regprop. I've also added a missing reginitcolors(), the lack of which was giving me stclass strings like "ANYOFR[(null)ab(null)]" and could presumably have caused coredumps elsewhere.

I'm happy to defer the question of including re_top.h to another time.

hvds committed:
fixup: missing aTHX_

hvds commented:
The only failed test this time is re/pat.t 955 under cygwin (timeout), which I believe is a known issue with our cygwin runs.

demerphq commented:
On Tue, 22 Sep 2020 at 00:01, Hugo van der Sanden
<notifications@github.com> wrote:
>
> The only failed test this time is re/pat.t 955 under cygwin (timeout), which I believe is a known issue with our cygwin runs.

I have one concern about this patch sequence. Generally i have put
functions like this in Hash::Util, not in re. The ones in re are the
bare minimum needed to do "i should be able to do this out of the box"
construction tasks with regex objects, whereas the ones in Hash::Util
are about introspecting a compiled hash object. So for instance we
have bucket_info(), and a handful of other introspection functions of
as similar nature to the one you are proposing.

So i think this patch should be moved to Hash::Util via ext/Hash-Util/Util.xs.

cheers,
Yves

hvds commented:
> I have one concern about this patch sequence. Generally i have put functions like this in Hash::Util, not in re. The ones in re are the bare minimum needed to do "i should be able to do this out of the box" construction tasks with regex objects, whereas the ones in Hash::Util are about introspecting a compiled hash object. So for instance we have bucket_info(), and a handful of other introspection functions of as similar nature to the one you are proposing.
>
> So i think this patch should be moved to Hash::Util via ext/Hash-Util/Util.xs.

I'm not sure what you mean by "functions like this", or "a compiled hash object" - it's a Regexp, not a hash. All the existing functions in Hash::Util have to do with hashes, not regexps. And ext/re already contains regmust() added via 256ddcd090 in 2006, which is a cut-down version of this new function. I think it is reasonable to say that the re module is our Regexp::Util.

TBH this would make more logical sense as a method on the Regexp class itself; but that raises back-compat risk, and we'd still need access to the debug-enabled functionality from ext/re/re_comp.c, so it would still need to live in re.xs.

demerphq commented:
Massive embarrassing brain fart. I posted that pre coffee and somehow
totally confused similar tasks with hashes and regexen in my head. My
apology.


Yves

On Tue, 22 Sep 2020 at 12:59, Hugo van der Sanden <notifications@github.com>
wrote:

> I have one concern about this patch sequence. Generally i have put
> functions like this in Hash::Util, not in re. The ones in re are the bare
> minimum needed to do "i should be able to do this out of the box"
> construction tasks with regex objects, whereas the ones in Hash::Util are
> about introspecting a compiled hash object. So for instance we have
> bucket_info(), and a handful of other introspection functions of as similar
> nature to the one you are proposing.
>
> So i think this patch should be moved to Hash::Util via
> ext/Hash-Util/Util.xs.
>
> I'm not sure what you mean by "functions like this", or "a compiled hash
> object" - it's a Regexp, not a hash. All the existing functions in
> Hash::Util have to do with hashes, not regexps. And ext/re already contains
> regmust() added via 256ddcd
> <https://github.com/Perl/perl5/commit/256ddcd0907fa1fc11538ea1a70ff79ba0167b40>
> in 2006, which is a cut-down version of this new function. I think it is
> reasonable to say that the re module is our Regexp::Util.
>
> TBH this would make more logical sense as a method on the Regexp class
> itself; but that raises back-compat risk, and we'd still need access to the
> debug-enabled functionality from ext/re/re_comp.c, so it would still need
> to live in re.xs.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Perl/perl5/pull/18145#issuecomment-696649395>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAZ5R3GPUB474ZBVB3CCMLSHB7P5ANCNFSM4RPBQH7A>
> .
>


-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

hvds commented:
> TBH this would make more logical sense as a method on the Regexp class itself; but that raises back-compat risk, and we'd still need access to the debug-enabled functionality from ext/re/re_comp.c, so it would still need to live in re.xs.

If we could do this and avoid the need for re.xs, that would have the big benefit (to people hacking regcomp.c, at least) of making it possible to run tests using this under miniperl.

Based on my earlier experiment, that would require removing the DEBUG_r() wrapping implicit in the DEBUG_TRIE_COMPILE_r in regprop, then removing the #ifdef DEBUGGING from around its guts.

I've also changed my mind about back-compat risk - I had envisaged the possibility of someone having inherited from Regexp and defined their own optimization method on the subclass, but that wouldn't be a problem at all.

I'd be happy to rewrite this PR if people think this is a valid option. It would allow the existing regmust() to be rewritten in pure perl as something like:
```
  sub regmust {
    my($qr) = shift;
    my $opt = $qr->optimization // return;
    return ($opt->{anchored} // $opt->{"anchored utf8"},
            $opt->{floating} // $opt->{"floating utf8"});
  }
```

demerphq commented:
On Tue, 22 Sep 2020 at 16:09, Hugo van der Sanden <notifications@github.com>
wrote:

> TBH this would make more logical sense as a method on the Regexp class
> itself; but that raises back-compat risk, and we'd still need access to the
> debug-enabled functionality from ext/re/re_comp.c, so it would still need
> to live in re.xs.
>
> If we could do this and avoid the need for re.xs, that would have the big
> benefit (to people hacking regcomp.c, at least) of making it possible to
> run tests using this under miniperl.
>
> Based on my earlier experiment, that would require removing the DEBUG_r()
> wrapping implicit in the DEBUG_TRIE_COMPILE_r in regprop, then removing the
> #ifdef DEBUGGING from around its guts.
>
> I've also changed my mind about back-compat risk - I had envisaged the
> possibility of someone having inherited from Regexp and defined their own
> optimization method on the subclass, but that wouldn't be a problem at all.
>
If this is at all related to my senile moment this morning then please dont
think about it more. I totally confused hashes and regexen.

Regarding the Regexp class, I view it as a vestigial remnant of a
past attempt at modeling regexpes that didnt work out very well and IMO in
an ideal world would go away entirely. In perl "compiled regex" is a *type*
not a *class*.  It is very much like a subroutine/code-ref wheras Regexp is
a class into which you can bless any ref type you like, iow, being member
of class Regexp does not make something a compiled regex, nor are all
compiled regexps a member or inheritor of class Regexp.

So I personally would not associate any form of regexp introspection
tooling that depends on a relationship to the Regexp class. Class and type
are orthogonal in perl as you know.

perl -Mre -le'my $qr=bless qr/foo/, "QR"; print $qr->isa("Regexp")||0;
print re::regmust($qr);'
0
foo

I'd be happy to rewrite this PR if people think this is a valid option. It
> would allow the existing regmust() to be rewritten in pure perl as
> something like:
>
Please dont on my account.


>   sub regmust {
>     my($qr) = shift;
>     my $opt = $qr->optimization // return;
>     return ($opt->{anchored} // $opt->{"anchored utf8"},
>             $opt->{floating} // $opt->{"floating utf8"});
>   }
>
>
I would expect this to use a function which takes a qr ref and returns an
object of a well known class that provides these methods. But IMO regmust
is nice as it is, no need to reimplement it I think.

cheers,
Yves

-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

hvds commented:
> If this is at all related to my senile moment this morning then please dont think about it more.

It wasn't, it was just an opportune moment to mention it. The point was: being able to run with miniperl a suite of tests of regexp optimization (which I intend to add) would substantially enhance the benefit of this function to me, and potentially also to the hordes of others that work on regcomp. (The regmust() example was just showing what it would look like, not something I was proposing to change.)

I don't think the Regexp-not-qr or qr-not-Regexp examples would be that big an issue in practice, but it is fair to argue that the category error could lead to confusion, and that's reason enough not to do it.

The benefit of miniperl remains though, so another possibility would be to keep it as a function called re::optimization, but still move the implementation to universal.c.

demerphq commented:
On Tue, 22 Sep 2020 at 20:18, Hugo van der Sanden <notifications@github.com>
wrote:

> If this is at all related to my senile moment this morning then please
> dont think about it more.
>
> It wasn't, it was just an opportune moment to mention it. The point was:
> being able to run with miniperl a suite of tests of regexp optimization
> (which I intend to add) would substantially enhance the benefit of this
> function to me, and potentially also to the hordes of others that work on
> regcomp. (The regmust() example was just showing what it would look like,
> not something I was proposing to change.)
>
Hordes? :-)


> I don't think the Regexp-not-qr or qr-not-Regexp examples would be that
> big an issue in practice, but it is fair to argue that the category error
> could lead to confusion, and that's reason enough not to do it.
>
FWIW, all I was saying is I wouldn't use Regexp. I dont see any problem
with having an oo api to regexp object introspection, I actually think its
a nice idea. I just think using Regexp for it might be problematic. Alas.
I might make it so that it "just worked" when the qr// was isa "Regexp",
but I would make it so that it was possible to construct the wrapper from a
qr that is not blessed directly into Regexp also.

Maybe supporting an api like my $re= Regexp->new($qr);

IOW, if the $qr isa Regexp return the $qr, otherwise return a scalar ref to
the QR. Then make methods check the reftype.  Something like that. It might
be easier to make this work if the actual wrapper class *isnt* Regexp, but
instead you could make Regexp inherit from the wrapper class.

The benefit of miniperl remains though, so another possibility would be to
> keep it as a function called re::optimization, but still move the
> implementation to universal.c.
>
I think its fine to move to something like this to universal.c if it makes
it usable from miniperl for testing.

Yves


-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

hvds commented:
I spent some more time trying to do that, but I've given up on that for now - I found that regprop relies on a lot more DEBUGGING-only code than just the one DEBUG_TRIE_COMPILE_r macro.

So I'd like to merge this PR as it is now, so we can start building tests on it, and come back later to the possibility of extending it to miniperl.

@xsawyerx is that ok?

xsawyerx commented:
> I spent some more time trying to do that, but I've given up on that for now - I found that regprop relies on a lot more DEBUGGING-only code than just the one DEBUG_TRIE_COMPILE_r macro.
> 
> So I'd like to merge this PR as it is now, so we can start building tests on it, and come back later to the possibility of extending it to miniperl.
> 
> @xsawyerx is that ok?

A-OK by me! :+1:

hvds commented:
Now merged with 7c253f2eee and 49781baac3.

hvds commented:
Ugh, now fixed with additional 54fc8aa1e9.

(I had a system crash while prepping the branch for merge; I thought I had recovered the prepped branch, but I had actually recovered to a SHA missing the aTHX_ on the my_regprop() call.)
